x-pack/plugins/apm/public/context/LocationContext.tsx

```diff
@@ -14,10 +14,18 @@ interface Props {
 const initialLocation = {} as Location;
 
 const LocationContext = createContext(initialLocation);
-const LocationProvider: React.FC<Props> = ({ history, ...props }) => {
+const LocationProvider: React.FC<Props> = ({ history, children }) => {
   const [location, setLocation] = useState(history.location);
-  history.listen(updatedLocation => setLocation(updatedLocation));
-  return <LocationContext.Provider {...props} value={location} />;
+
+  useEffect(() => {
+    const unlisten = history.listen(updatedLocation => {
+      setLocation(updatedLocation);
+    });
+
+    return unlisten;
+  }, []);
```

Comment:
So putting this in a `useEffect` (with empty deps) prevents it from re-subscribing on every render of this Provider, right? I get that. What is the effect of the `unlisten` callback and how does re-subscribing again "un" listen?

Reply:
Ah nevermind, I see that it returns the unsubscribe function from `history.listen`. Makes sense now :+1:

Reply:
> So putting this in a useEffect (with empty deps) prevents it from re-subscribing on every render of this Provider, right? 

Yes, any side effects should happen in `useEffect` - never in the (render) function. 

> What is the effect of the unlisten callback and how does re-subscribing again "un" listen?

To avoid memory leaks event listeners should always be removed after use. Even if this is only invoked once, it should still be removed when navigating away from APM. Else switching between APM and another plugin would produce more and more listeners.
Kibana currently does full page refreshes so probably not going to make a difference - but if we fix that this is important.

https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect

Reply:
(Didn't see your second reply until I posted... GH is too slow...)